plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp

```diff
@@ -0,0 +1,172 @@
+using chain::packed_transaction;
+using chain::transaction_id_type;
+
+using snapshot_request_information = producer_plugin::snapshot_request_information;
+using get_snapshot_requests_result = producer_plugin::get_snapshot_requests_result;
+
+class snapshot_scheduler_listener {
+   virtual void on_block(uint32_t height) = 0;
+   virtual void on_irreversible_block(uint32_t height) = 0;
+   virtual void on_abort_block() = 0;
+};
+
+class snapshot_scheduler_handler {
+   virtual void schedule_snapshot(const snapshot_request_information& sri) = 0;
+   virtual void unschedule_snapshot(uint32_t sri) = 0;
+   virtual get_snapshot_requests_result get_snapshot_requests() = 0;
+};
+
+class snapshot_scheduler : public snapshot_scheduler_handler, public snapshot_scheduler_listener {
+private:
+   struct by_snapshot_id;
+   struct by_snapshot_value;
+   struct as_vector;
+
+   using snapshot_requests = bmi::multi_index_container<
+         snapshot_request_information,
+         indexed_by<
+               bmi::hashed_unique<tag<by_snapshot_id>, BOOST_MULTI_INDEX_MEMBER(snapshot_request_information, uint32_t, snapshot_request_id)>,
+               bmi::random_access<tag<as_vector>>,
+               bmi::ordered_unique<tag<by_snapshot_value>,
+                                   composite_key<snapshot_request_information,
+                                                 BOOST_MULTI_INDEX_MEMBER(snapshot_request_information, uint32_t, block_spacing),
+                                                 BOOST_MULTI_INDEX_MEMBER(snapshot_request_information, uint32_t, start_block_num),
+                                                 BOOST_MULTI_INDEX_MEMBER(snapshot_request_information, uint32_t, end_block_num)>>>>;
+   snapshot_requests _snapshot_requests;
+   snapshot_db_json _snapshot_db;
+   uint32_t _snapshot_id = 0;
+   std::function<void(producer_plugin::next_function<producer_plugin::snapshot_information>)> _create_snapshot;
+
+public:
+   snapshot_scheduler() {}
+
+   // snapshot_scheduler_listener
+   virtual void on_block(uint32_t height) {
+      for(const auto& req: _snapshot_requests.get<0>()) {
+         // execute "asap" or if matches spacing
+         if((req.start_block_num == 0) ||
+            (!((height - req.start_block_num) % req.block_spacing))) {
+            execute_snapshot();
+         }
+         // assume "asap" for snapshot with missed/zero start, it can have spacing
+         if(req.start_block_num == 0) {
+            auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+            auto it = snapshot_by_id.find(req.snapshot_request_id);
+            _snapshot_requests.modify(it, [&height](auto& p) { p.start_block_num = height; });
+         }
+         // remove expired request
+         if(req.end_block_num > 0 && height >= req.end_block_num) {
+            auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+            _snapshot_requests.erase(snapshot_by_id.find(req.snapshot_request_id));
```

Comment:
I think this can just be `_snapshot_requests.erase(req.snapshot_request_id);`

Reply:
yes, that is better :) done